### ISSUE TITLE
Target loss

### ISSUE THREAD

**Issue by Labaien96** [facebookresearch/detr#78]
Hi, I'm working with a custom dataset, which contains images with small defects, which are annotated with bounding boxes. The goal in my case is to detect these bugs, so I have set the number of classes to 2, one for defect, and the other one non-defect.  The targets dicts contain the key "labels" containing a tensor of dim `[nb_target_boxes]`, which I have created as `torch.ones((num_objs,), dtype=torch.int64)`. 
```
    def loss_labels(self, outputs, targets, indices, num_boxes, log=True):
        """Classification loss (NLL)
        targets dicts must contain the key "labels" containing a tensor of dim [nb_target_boxes]
        """
        assert 'pred_logits' in outputs
        src_logits = outputs['pred_logits']

        idx = self._get_src_permutation_idx(indices)
        target_classes_o = torch.cat([t["labels"][J] for t, (_, J) in zip(targets, indices)])
        target_classes = torch.full(src_logits.shape[:2], self.num_classes,
                                    dtype=torch.int64, device=src_logits.device)
        target_classes[idx] = target_classes_o

        loss_ce = F.cross_entropy(src_logits.transpose(1, 2), target_classes, self.empty_weight)
        losses = {'loss_ce': loss_ce}

        if log:
            # TODO this should probably be a separate loss, not hacked in this one here
            losses['class_error'] = 100 - accuracy(src_logits[idx], target_classes_o)[0]
        return losses
```

I'm trying to understand how the loss of labels is computed but I don't understand certain things about the loss_labels function. I have several doubts:

1. First of all, I would like to know if the way I pass the labels is the right way for a two-class problem and what I should put in num_classes, 1 or 2. There are only bounding boxes around a defect, so I pass only these labels, as I explained above. 
2. In the code above, src_logits why does it give me a dimension output [batch_size,num_queries,num_classes+1]? why +1? 
3. target_classes_o: I don't understand exactly what this is, is it the ground truth?
4. target_classes: I understand that here first we have all as a non-defect class and in the indexes that correspond to defect I change the target, but it is not clear to me.
5. What is the function of empty_weight when calculating the loss? 

6. If I am working with only two classes, should I put binary_cross_entropy instead of cross_entropy?

Thank you in advance.

**Comment by alcinos** [comment]
Hi @Labaien96 
No matter the number of classes you have, DETR will add an additional class corresponding to "no-object". 
As for your questions:

1. num_classes should be equal to the maximal class label + 1. So if your labels are 0 and 1, you need to use num_classes = 2
2. The +1 here correspond to the no-object class (DETR reserves the last class id for it)
3. `target_classes_o` are the ground truth labels for the existing objects in the image, in order defined by the hungarian matcher
4. Since DETR predicts always 100 objects, if there are `n` "true" objects in the image (ie `len(target_classes_o) = n`, we must supervise the `100 - n` remaining outputs to predict "no-object". See that `target_classes` is initialized with 100 "no-objects" labels, and then we fill the "true" objects with their correct label coming from `target_classes_o`
5. On average, the number of true objects on the image is much smaller than the total number of predicted objects (100), so the "no-object" class is the majority. `empty_weight` is used to rebalance the classification loss and diminish the relative weight of the "no-object" class.
6. Since DETR also predicts the "no-object" class, you'll effectively have 3 classes, thus you need to keep the cross_entropy as is.

Generally speaking, I don't think your use-case warrants any modification of the losses as they are defined now, beyond setting the correct `num_classes` (although of course understanding how they work is always valuable)
If you want to learn more about the "no-object" and the hungarian loss, I recommend reading our paper.

I believe I have answered your questions, and as such I'm closing this. Feel free to reach out if you have further concerns.

**Comment by Labaien96** [comment]
> Hi @Labaien96
> No matter the number of classes you have, DETR will add an additional class corresponding to "no-object".
> As for your questions:
> 
> 1. num_classes should be equal to the maximal class label + 1. So if your labels are 0 and 1, you need to use num_classes = 2
> 2. The +1 here correspond to the no-object class (DETR reserves the last class id for it)
> 3. `target_classes_o` are the ground truth labels for the existing objects in the image, in order defined by the hungarian matcher
> 4. Since DETR predicts always 100 objects, if there are `n` "true" objects in the image (ie `len(target_classes_o) = n`, we must supervise the `100 - n` remaining outputs to predict "no-object". See that `target_classes` is initialized with 100 "no-objects" labels, and then we fill the "true" objects with their correct label coming from `target_classes_o`
> 5. On average, the number of true objects on the image is much smaller than the total number of predicted objects (100), so the "no-object" class is the majority. `empty_weight` is used to rebalance the classification loss and diminish the relative weight of the "no-object" class.
> 6. Since DETR also predicts the "no-object" class, you'll effectively have 3 classes, thus you need to keep the cross_entropy as is.
> 
> Generally speaking, I don't think your use-case warrants any modification of the losses as they are defined now, beyond setting the correct `num_classes` (although of course understanding how they work is always valuable)
> If you want to learn more about the "no-object" and the hungarian loss, I recommend reading our paper.
> 
> I believe I have answered your questions, and as such I'm closing this. Feel free to reach out if you have further concerns.

Thank you @alcinos, I really appreciate your response.  The thing is that in my case I'm only trying to detect defects, so I only have one class (2 if I add the "no object" class), so if an image has only one bounding box, the label I give it is [1], if it has two bounding boxes the label I give it is [1,1], etc. That is, the dataset that I have, passes the targets as many "ones" as bounding boxes has the image, so at the beginning should I put `target_classes` with all zeros and fill the indexes that have objects with ones?

Thank you in advance.

**Comment by alcinos** [comment]
So if you decide to use label = 1 for the things you want to detect you need to:

- use num_classes = 2
- For each image, you need to have a target dictionnary with the keys "labels" (which in your case will be a vector of 1s) and "boxes" which contains as many boxes as you have elements in the label vector, with format (center_x, center_y, h, w), normalized by the image size.
- You don't need to modify anything from the losses, and you don't need to care about "no-objects", they will be handled directly by DETR.

Hope this helps.

**Comment by Labaien96** [comment]
Thank you @alcinos ! Of course it is helpful! Despite that, `loss_giou` doesn't go down much. I think it may be because of the size of the bounding boxes in my dataset, as the bounding boxes I have are very small (because the defects are small) compared to what the model predicts. Why doesn't the last layer of the MLP you use to predict the bounding boxes have activation? I tried putting a sigmoid in the last layer but the results are still bad. Another reason why I think it is failing could be in the way I normalize, since I used the same parameters as the one you use for COCO, what do you think?

**Comment by alcinos** [comment]
We do have a sigmoid for the box prediction already:
https://github.com/facebookresearch/detr/blob/22d955b8bef32003ee2aa065d4a8fa0ea1869680/models/detr.py#L68
So it's not necessary to add one more.

Can you define what you mean by "loss_giou doesn't go down much"? Bear in mind that DETR is typically trained for a very long time compared to "traditional" detectors, so you may want to wait a bit before drawing conclusions.
Small objects are not DETR's main strength, especially if the bounding boxes are smaller than 32x32.

**Comment by Labaien96** [comment]
Hi @alcinos, I didn't realize about the sigmoid, thank you. Oh, okay, in my case the objects are smaller than 16x16 pixels, so I understand that DETR won't work well...
